Fix some SSL errors in mail.py causing SMTP Syntax Errors (Rebase of https://github.com/ansible/ansible-modules-extras/pull/708 )

abadger opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
lib/ansible/modules/notification/mail.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
2.3
```

##### SUMMARY

708 was full of extraneous merge commits interwoven with commits to
implement the feature. In the end the only way I could clean this up
in reasonable time was to just take a regular diff between the PR and
the base.  This lost the history of intermediate commits but I've
preserved attribution to @dayton967 via git's --author field.

Although I preserved the logic of the PR, there were a few additional
things that I cleaned up:

* Fixed import of email.mime.multipart
* Used the argspec to set port and timeout to integers instead of having
  ad hoc code inside of the module.
* Used argspec's choices for secure instead of ad hoc code inside of the
  module.
* Removed some unused variables
* Made secure_state a python boolean instead of using 0 and 1
* Used secure with string comparisons instead of turning it into an
  integer code.  This is much more readable.
* Fixed catching of SMTPExceptions (SMTPException wasn't imported
  directly so it needed to use the smtplib namespace.)

dayton967 committed:
Rebase of https://github.com/ansible/ansible-modules-extras/pull/708

708 was full of extraneous merge commits interwoven with commits to
implement the feature. In the end the only way I could clean this up
in reasonable time was to just take a regular diff between the PR and
the base.  This lost the history of intermediate commits but I've
preserved attribution to @dayton967 via git's --author field.

Although I preserved the logic of the PR, there were a few additional
things that I cleaned up:

* Fixed import of email.mime.multipart
* Used the argspec to set port and timeout to integers instead of having
  ad hoc code inside of the module.
* Used argspec's choices for secure instead of ad hoc code inside of the
  module.
* Removed some unused variables
* Made secure_state a python boolean instead of using 0 and 1
* Used secure with string comparisons instead of turning it into an
  integer code.  This is much more readable.
* Fixed catching of SMTPExceptions (SMTPException wasn't imported
  directly so it needed to use the smtplib namespace.)

abadger commented:
@dayton967 @dagwieers @troy2914

Okay, here's the "rebase" of https://github.com/ansible/ansible-modules-extras/pull/708  Please check it to make sure I didn't mess anything up.

When I cleaned up the code I was able to read it a bit easier and came up with a few questions but I don't know enough about the smtplib code to answer them for myself:

* Why does the code default value ```Line 266:    code = 000``` and the conditional value ```Line 300: if int(code) > 000:``` use ```000```?  Should it just be 0 or is there a non-zero digit missing?
* Why doesn't always try to upgrade the connection via starttls?  I don't know if there's some insecurity around the difference between what always does and what starttls does but it seems like a user would only care that one of those two had established an encrypted connection.  With the current parameters, the user would have to know which of the two to use. Maybe always and starttls should be merged into a single secure strategy?

dagwieers commented:
- I guess it states `000` because SMTP typically works with 3 digits status codes, but it should not matter.
- We had the discussion and there  were 2 views on this, one was that it would be better to have a secure connection as much as possible when possible, the other view was that if the user explicitly does not want STARTTLS, (e.g. for debugging or whatever other reason) the module shouldn't try to be smart. I am fine with both views, but if the default is to be as secure as possible, then I don't mind if people can also opt-in for a less secure options.
shipit :-)

abadger committed:
Change code from 000 to 0

To avoid confusion.  The former leads to questions:
Is it missing a digit?  Is it supposed to be octal?

abadger commented:
Merged to devel for 2.3 release. Thanks @dayton967 for the fix and @dagwieers and @troy2914 for the review.
